docs(near-sdk-sim): add lots of detail to README (#327)
near/near-sdk-rs: near-sdk-sim/README.md

# NEAR Simulator & cross-contract testing library

When writing NEAR contracts, with Rust or other Wasm-compiled languages like [AssemblyScript](https://github.com/near/near-sdk-as), the default testing approach for your language of choice (such as `mod test` in your Rust project's `src/lib.rs` file) is great for testing the behavior of a single contract in isolation.

> But the true power of blockchains & smart contracts comes from cross-contract calls. How do you make sure your cross-contract code works as you expect?

Reviewer: ```suggestion
But the true power of blockchains and smart contracts comes from cross-contract calls. How do you make sure your cross-contract code works as you expect?
```

Author: y tho

As a first step, you can use this library! With it, you can:

* Test cross-contract calls

> * Accurately check [gas](https://docs.near.org/docs/concepts/gas) & [storage](https://docs.near.org/docs/concepts/storage-staking) usage for your contract

Reviewer: This isn't completely true, but you can profile the gas usage to understand where your costs are going.  We found that the gas used on testnet was higher so this does provide a lower bound on gas usage.

Author: How about I change to:

```suggestion
* Profile [gas](https://docs.near.org/docs/concepts/gas) & [storage](https://docs.near.org/docs/concepts/storage-staking) usage for your contract, establishing lower bounds for costs of deployed contracts and rapidly identifying problematic areas prior to deploying.
```

    // Use `None` for default genesis configuration
    let root = init_simulator(None);

    let contract = root.create_user(
        CONTRACT_ID.to_string(), // account id
        to_yocto("100"),         // initial balance
    );

    let contract = contract.deploy(
        &CONTRACT_WASM_BYTES,
        CONTRACT_ID.to_string(),
        STORAGE_AMOUNT
    );

    (root, contract)
}
```

Now you can start writing tests:

```rust
use near_sdk::serde_json::json;
use near_sdk_sim::DEFAULT_GAS;

#[test]
fn simulate_some_view_function() {
    let (root, contract) = init();

    let actual: String = root.view(
        CONTRACT_ID.to_string(),
        "view_something".to_string(),
        json!({
            "some_param": "some_value".to_string(),
        }).to_string().into_bytes()
    ).unwrap_json();

    assert_eq!("expected", actual);
}

#[test]
fn simulate_some_change_method() {
    let (root, contract) = init();

    let result = root.call(
        CONTRACT_ID.to_string(),
        "change_something".to_string(),
        json!({
            "some_param": "some_value".to_string(),
        }).to_string().into_bytes()
        DEFAULT_GAS / 2,
        0 // attached deposit
    );

>     assert!(result.is_ok());

Author: @willemneal will this & the view call above work properly once https://github.com/near/near-sdk-rs/pull/325 is merged, or did I mess up some syntax?

Reviewer: Correct, this should work with the PR mentioned.  We could also consider adding the README to the doc tests: https://doc.rust-lang.org/rustdoc/documentation-tests.html#include-items-only-when-collecting-doctests

}
```

## Optional macros

The above approach is a good start, and will work even if your Wasm files are compiled from a language other than Rust.

But if your original files are Rust and you want better ergonomics while testing, `near-sdk-sim` provides a nice bonus feature.

> `near-sdk-sim` modifies the `near_bindgen` macro from `near-sdk` to create an additional struct+implementation from your contract, with `Contract` added to the end of the name, like `xxxxxContract`. So if you have a contract with `[package].name` set to `token` with this in its `src/lib.rs`:

Reviewer: ```suggestion
`near-sdk-sim` modifies the `near_bindgen` macro from `near-sdk` to create an additional struct and implementation from your contract, with `Contract` added to the end of the name, like `xxxxxContract`. So if you have a contract with `[package].name` set to `token` with this in its `src/lib.rs`:
```

Reviewer: You could have a struct for just the data, and the implementation is optional.  In this case the data that the new struct has is the contract's account_id so that when you call methods it knows the receiver.

```rust
#[near_bindgen]
struct Token {
    ...
}

#[near_bindgen]
impl Token {
    ...
}
```

Then in your simulation test you can import `TokenContract`:

```rust
use token::TokenContract;
```

Now you can simplify the `init` & test code from the previous section:

> ```

Reviewer: ```suggestion
```

```rust
use near_sdk::serde_json::json;
use near_sdk_sim::{DEFAULT_GAS, STORAGE_AMOUNT, call, deploy, init_simulator, to_yocto, view};

fn init() {
    let root = init_simulator(None);

    let contract = deploy!(
        contract: TokenContract,
        contract_id: CONTRACT_ID,
        bytes: &CONTRACT_WASM_BYTES,
        signer_account: root,
    );

    (root, contract)
}

#[test]
fn simulate_some_view_function() {
    let (root, contract) = init();

    let actual: String = view!(
        contract.view_something("some_value".to_string()),
    ).unwrap_json();

    assert_eq!("expected", actual);
}

#[test]
fn simulate_some_change_method() {
    let (root, contract) = init();

    let result = call!(

>         contract.change_something("some_value".to_string()),

Reviewer: ```suggestion
        root,
        contract.change_something("some_value".to_string()),
```
Missing the signer.

        bytes: &CONTRACT_WASM_BYTES,
        signer_account: root,
    );

    (root, contract)
}

#[test]
fn simulate_some_view_function() {
    let (root, contract) = init();

    let actual: String = view!(
        contract.view_something("some_value".to_string()),
    ).unwrap_json();

    assert_eq!("expected", actual);
}

#[test]
fn simulate_some_change_method() {
    let (root, contract) = init();

    let result = call!(
        root,
        contract.change_something("some_value".to_string()),
        gas = DEFAULT_GAS / 2,
        deposit = 0
    );

    assert!(result.is_ok());
}
```

# Common patterns

## Profile gas costs

TODO

## Profile storage costs

TODO

## Inspect intermediate state of all calls in a complicated chain of transactions 

TODO

## Check expected transaction failures

# Tweaking the genesis config

For many simulation tests, using `init_simulator(None)` is good enough. This uses the [default genesis configuration settings](https://github.com/near/near-sdk-rs/blob/0a9a56f1590e1f19efc974160c88f32efcb91ef4/near-sdk-sim/src/runtime.rs#L59-L72).

## Why would you want to change these?

TODO

## How do you change them?

> TODO

Author: @willemneal do you mind filling in this "Tweaking the genesis config" section? It doesn't need to be long.

```

Note that you need to add the `tag` (or `commit`) of the version.

## Workspace setup

If you want to check gas & storage usage of one Rust contract, you can add the above dependencies to `Cargo.toml` in the root of your project. If you want to test cross-contract calls, we recommend setting up a cargo [workspace](https://doc.rust-lang.org/cargo/reference/workspaces.html). Here's how it works:

Let's say you have an existing contract project in a folder called `contract`.

Go ahead and make a subfolder within it called `contract`, and move the original contents of `contract` into this subfolder. Now you'll have `contract/contract`. You can rename the root folder to something like `contracts` or `contract-wrap`, if you want. Some bash commands to do this:

```bash
mkdir contract-wrap
mv contract contract-wrap
```

Now in the root of the project (`contract-wrap`), create a new `Cargo.toml`. You'll have to add the normal `[package]` section, but unlike most projects you won't have any `dependencies`, only `dev-dependencies` and a `workspace`:

```toml
[dev-dependencies]
near-sdk = { git = "https://github.com/near/near-sdk-rs.git", tag="3.0.0" }

> near-sdk-sim = { git = "https://github.com/near/near-sdk-rs.git", tag="3.0.0" }

Reviewer 2: What's the plan to update it? Pre-publishing?

Author: I spoke with @willemneal and he thinks it makes sense to publish the package with this, then make a small doc update right after.

use near_sdk_sim::{call, view};
use crate::utils::init;

#[test]
fn simulate_some_view_function() {
    let (root, contract, _alice) = init();

    let actual: String = view!(
        contract.view_something("some_value".to_string()),
    ).unwrap_json();

    assert_eq!("expected", actual);
}

#[test]
fn simulate_some_change_method() {
    let (root, contract, _alice) = init();

    // uses default gas amount
    let result = call!(
        root,
        contract.change_something("some_value".to_string()),
        deposit = 1,
    );

    assert!(result.is_ok());
}
```

# Common patterns

## Profile gas costs

For a chain of transactions kicked off by `call` or `call!`, you can check the `gas_burnt` and `tokens_burnt`, where `tokens_burnt` will equal `gas_burnt` multiplied by the `gas_price` set in the genesis config. You can also print out `profile_data` to see an in-depth gas-use breakdown.

```rust
let outcome = some_account.call(
    "some_contract",
    "method",
    &json({
        "some_param": "some_value",
    }).to_string().into_bytes(),
    DEFAULT_GAS,
    0,
);

println!(
    "profile_data: {:#?} \n\ntokens_burnt: {}Ⓝ",
    outcome.profile_data(),

>     (outcome.tokens_burnt()) as f64 / f64::powi(10.0, 24)

Reviewer 2: ```suggestion
    (outcome.tokens_burnt()) as f64 / 1e24f64
```

Author: Thanks, `1e24` works!

Checking this revealed that `profile_data` is now broken. I created https://github.com/near/near-sdk-rs/pull/339 to track this issue.
